x/auth/tx/service.go

```diff
@@ -157,6 +158,65 @@ func (s txServer) GetTx(ctx context.Context, req *txtypes.GetTxRequest) (*txtype
 	}, nil
 }
 
+// protoTxProvider is a type which can provide a proto transaction. It is a
+// workaround to get access to the wrapper TxBuilder's method GetProtoTx().
+type protoTxProvider interface {
+	GetProtoTx() *txtypes.Tx
+}
+
+// GetBlockWithTxs returns a block with decoded txs.
+func (s txServer) GetBlockWithTxs(ctx context.Context, req *txtypes.GetBlockWithTxsRequest) (*txtypes.GetBlockWithTxsResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "request cannot be nil")
+	}
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	currentHeight := sdkCtx.BlockHeight()
+
+	if req.Height < 1 || req.Height > currentHeight {
+		return nil, sdkerrors.ErrInvalidHeight.Wrapf("requested height %d but height must not be less than 1 "+
+			"or greater than the current height %d", req.Height, currentHeight)
+	}
+
+	node, err := s.clientCtx.GetNode()
+	if err != nil {
+		return nil, err
+	}
+
+	blockRes, err := node.Block(ctx, &req.Height)
+	if err != nil {
+		return nil, err
+	}
+	block := blockRes.Block
+	blockId := blockRes.BlockID
+
+	txs := make([]*txtypes.Tx, len(block.Txs))
+	for i, tx := range block.Txs {
+		txb, err := s.clientCtx.TxConfig.TxDecoder()(tx)
+		if err != nil {
+			return nil, err
+		}
+		p, ok := txb.(protoTxProvider)
+		if !ok {
+			return nil, fmt.Errorf("could not cast %T to %T", txb, txtypes.Tx{})
+		}
+		txs[i] = p.GetProtoTx()
+	}
+
+	protoBlockId := blockId.ToProto()
+	protoBlock, err := block.ToProto()
```

Comment:
there's some duplicate code between this and client/grpc/tmservice's GetBlockByHeight. Do you think it's worth/easy to factor both into a function?

Reply:
sure, added a function for this